### PR TITLE
ARROW-4439: [C++] Improve FindBrotli.cmake

### PR DESCRIPTION
**Pull request by rip-nsk**

### CODE REVIEW

**Comment by wesm** [comment]
rebased

**Comment by rip-nsk** [comment]
WIP

**Comment by xhochy** [comment]
Rebased.

**Comment by wesm** [comment]
I just fixed the Linux build. Could you write in the PR description what you are trying to accomplish with this patch?

**Comment by wesm** [comment]
Looks like this also broke library detection in the manylinux1 build. I don't have time to fix it right now

**Comment by rip-nsk** [comment]
> I just fixed the Linux build. Could you write in the PR description what you are trying to accomplish with this patch?

The goal of this PR find libraries with respect to BROTLI_STATIC_LIB_SUFFIX/BROTLI_MSVC_STATIC_LIB_SUFFIX (BTW the can be "merged") for case when only static library is available

Difference from previous logic: BROTLI_STATIC_LIB_SUFFIX should be explicitly specified,  there are no auto select between "-static" and "" anymore.

**Comment by rip-nsk** [comment]
I don't know how to pass empty BROTLI_STATIC_LIB_SUFFIX to fix manylinux1 build

File: cpp/cmake_modules/FindBrotli.cmake [hunk 1]
```diff
@@ -43,66 +42,53 @@ find_path( BROTLI_INCLUDE_DIR NAMES brotli/decode.h
...
-      ${BROTLI_LIBS}/${CMAKE_SHARED_LIBRARY_PREFIX}${BROTLI_LIB_NAME}enc${CMAKE_SHARED_LIBRARY_SUFFIX}
-      ${BROTLI_LIBS}/${CMAKE_SHARED_LIBRARY_PREFIX}${BROTLI_LIB_NAME}dec${CMAKE_SHARED_LIBRARY_SUFFIX}
-      ${BROTLI_LIBS}/${CMAKE_SHARED_LIBRARY_PREFIX}${BROTLI_LIB_NAME}common${CMAKE_SHARED_LIBRARY_SUFFIX})
-else ()
-  set(BROTLI_FOUND FALSE)
-endif ()
+  find_library(BROTLI_SHARED_${LIBRARY}
+    NAMES ${CMAKE_SHARED_LIBRARY_PREFIX}${BROTLI_LIB_NAME}${BROTLI_LIB}${BROTLI_SHARED_LIB_SUFFIX}${CMAKE_SHARED_LIBRARY_SUFFIX}
+    PATHS ${_brotli_roots} NO_DEFAULT_PATH
+    PATH_SUFFIXES ${BROTLI_PATH_SUFFIXES})
```
Comment (xhochy) [inline cpp/cmake_modules/FindBrotli.cmake]: I'm wondering whether we should simply remove `NO_DEFAULT_PATH` in some cases. Then `cmake` would be able to find the libraries in the system default paths.

Comment (rip-nsk) [reply cpp/cmake_modules/FindBrotli.cmake]: it works on my centos7:
-- Found the Brotli static library: /usr/lib64/libbrotlienc.a;/usr/lib64/libbrotlidec.a;/usr/lib64/libbrotlicommon.a

**Comment by rip-nsk** [comment]
@wesm, @xhochy 
Any reasons blocks this PR?

**Comment by wesm** [comment]
The big CMake refactor patch is blocking this. Can you rebase after that goes in?

**Comment by wesm** [comment]
Can you also write a PR description explaining what issue is being fixed?

**Comment by xhochy** [comment]
This patch is probably already digested in the things I did in https://github.com/apache/arrow/pull/3688 @rip-nsk Can you have a look there if I have covered all cases?

**Comment by rip-nsk** [comment]
> This patch is probably already digested in the things I did in #3688 @rip-nsk Can you have a look there if I have covered all cases?
As far I can see, #3688 doesn't use BROTLI_STATIC_LIB_SUFFIX, so if it is not equal to some known values (""|"-static"|"_static") library will not be found.
Also, it seems mixes static/shared libraries.

**Comment by wesm** [comment]
Can you have a look now that the CMake refactor has been merged?

**Comment by wesm** [comment]
I'm removing this from the 0.13 milestone and moving to 0.14. If you are able to fix this this week we can have a closer look. Please add a PR description for the changelog (there is not description on JIRA either)

**Comment by emkornfield** [comment]
@rip-nsk do you think you will be able to rebase this to get it reviewed?

**Comment by wesm** [comment]
ping @rip-nsk

**Comment by rip-nsk** [comment]
It is hard to "resolve conflicts" with base which is changed so much..
I'll recreate this PR later (not time for now)

**Pull request closed without merging**